src/tree/updater_gpu_hist.cu

```diff
@@ -348,41 +330,50 @@ struct DeviceShard {
 
   dh::CubMemory temp_memory;
 
+  // TODO(canonizer): do add support multi-batch DMatrix here
   DeviceShard(int device_idx, int normalised_device_idx,
-              bst_uint row_begin, bst_uint row_end, int n_bins, TrainParam param)
+              bst_uint row_begin, bst_uint row_end, TrainParam param)
     : device_idx(device_idx),
       normalised_device_idx(normalised_device_idx),
       row_begin_idx(row_begin),
       row_end_idx(row_end),
+      row_stride(0),
       n_rows(row_end - row_begin),
-      n_bins(n_bins),
-      null_gidx_value(n_bins),
+      n_bins(0),
+      null_gidx_value(0),
       param(param),
       prediction_cache_initialised(false),
       can_use_smem_atomics(false) {}
 
-  void Init(const common::HistCutMatrix& hmat, const SparsePage& row_batch) {
-    // copy cuts to the GPU
+  void InitRowPtrs(const SparsePage& row_batch) {
     dh::safe_cuda(cudaSetDevice(device_idx));
-    thrust::device_vector<float> cuts_d(hmat.cut);
-    thrust::device_vector<size_t> cut_row_ptrs_d(hmat.row_ptr);
-
-    // find the maximum row size
-    thrust::device_vector<size_t> row_ptr_d(
-        row_batch.offset.data() + row_begin_idx, row_batch.offset.data() + row_end_idx  + 1);
-
-    auto row_iter = row_ptr_d.begin();
+    row_ptrs.resize(n_rows + 1);
+    thrust::copy(&row_batch.offset[row_begin_idx], &row_batch.offset[row_end_idx + 1],
```

Comment:
`&row_batch.offset[row_end_idx + 1]`

This is actually an array out of bounds error if the compiler does not optimise. MSVC will crash in debug mode on the statement. Instead use:
`row_batch.offset.data() + row_end_idx  + 1`

Also regarding thrust copy, I ran into issues with this creating large temporary memory buffers when copying nontrivial types (GradientPair). It should be fine for primitive types but I was considering banning thrust::copy completely via some kind of lint rule. Any thoughts on this?